modmail/extensions/threads.py

```diff
@@ -0,0 +1,459 @@
+            await ticket.recipient.create_dm()
+        if message.guild is not None:
+            # thread -> dm
+            logger.debug(
+                "Relaying message id {0} by {3} from thread {1} to dm channel {2}.".format(
+                    message.id, ticket.thread.id, ticket.recipient.dm_channel.id, message.author
+                )
+            )
+            sent_message = await ticket.recipient.send(
+                embed=ticket.embed_creator.create_message_embed_to_user(message, contents)
+            )
+            # also relay it in the thread channel
+            new_message = await ticket.thread.send(
+                embed=ticket.embed_creator.create_message_embed_to_user(message, contents)
+            )
+            await message.delete()
+            message = new_message
+            ticket.last_sent_message = message
+        else:
+            # dm -> thread
+            logger.debug(
+                "Relaying message id {0} from dm channel {1} with {3} to thread {2}.".format(
+                    message.id, ticket.recipient.dm_channel.id, ticket.thread.id, message.author
+                )
+            )
+            sent_message = await ticket.thread.send(
+                embed=ticket.embed_creator.create_message_embed_to_guild(message)
+            )
+
+        # add messages to the dict
+        ticket.messages[message] = sent_message
+        return sent_message
+
+    @is_modmail_thread()
+    @commands.command(aliases=("r",))
+    async def reply(self, ctx: Context, *, message: str) -> None:
+        """Send a reply to the user."""
+        ticket = self.get_ticket(ctx.channel.id)
+        await self._relay_message(ticket, ctx.message, message)
+
+    @is_modmail_thread()
+    @commands.command(aliases=("e",))
+    async def edit(self, ctx: Context, message: Optional[discord.Message] = None, *, content: str) -> None:
+        """Edit a message in the thread."""
+        ticket = self.get_ticket(ctx.channel.id)
+        if message is None:
+            message = ticket.last_sent_message
+        await ticket.messages[message].edit(
+            embed=ticket.embed_creator.create_edited_message_embed_to_user(content, message)
+        )
+        await ticket.thread.send(
+            embed=ticket.embed_creator.create_edited_message_embed_to_guild(content, message)
+        )
+        await ctx.message.add_reaction(ON_SUCCESS_EMOJI)
+
+    @is_modmail_thread()
+    @commands.command(aliases=("d", "del"))
+    async def delete(self, ctx: Context, message: Optional[discord.Message] = None) -> None:
+        """Delete a message in the thread."""
+        ticket = self.get_ticket(ctx.channel.id)
+        if message is None:
+            message = ticket.last_sent_message
+            ticket.last_sent_message = None
+        await ticket.messages[message].delete()
+        await message.delete()
+        await ctx.message.add_reaction(ON_SUCCESS_EMOJI)
+
+    async def _close_thread(
```

Comment:
Move this above the "commands" region.

Reply:
see above https://github.com/discord-modmail/modmail/pull/53#discussion_r721031565